nautobot/extras/groups.py

```diff
@@ -0,0 +1,182 @@
+import logging
+import urllib
+
+from django import forms
+from django.db import models
+from django.urls import reverse
+from django.utils.functional import classproperty
+
+from nautobot.utilities.utils import get_filterset_for_model, get_form_for_model, get_route_for_model
+
+
+logger = logging.getLogger(__name__)
+
+
+def dynamicgroup_map_factory(model):
+    """Generate a `FooDynamicGroupMap` class for a given `model`."""
+
+    try:
+        if not issubclass(model, models.Model):
+            return None
+    except TypeError:
+        return None
+
+    filterset_class = get_filterset_for_model(model)
+    filterform_class = get_form_for_model(model, form_prefix="Filter")
+    form_class = get_form_for_model(model)
+
+    group_map = type(
+        str("%sDynamicGroupMap" % model._meta.object_name),
+        (BaseDynamicGroupMap,),
+        {
+            "model": model,
+            "filterset_class": filterset_class,
+            "filterform_class": filterform_class,
+            "form_class": form_class,
+        },
+    )
+
+    return group_map
+
+
+class BaseDynamicGroupMap:
+    """
+    Dynamic Group mapping used to generate mappings for each model class.
+
+    This class itself should not be invoked directly as the class variables will
+    not be populated and most class methods will fail.
+    """
+
+    model = None
+    form_class = None
+    filterset_class = None
+    filterform = None
+
+    # This is used as a `startswith` check on field names, so these can be
+    # explicit fields or just substrings.
+    #
+    # Currently this means skipping "search", custom fields, and custom relationships.
+    #
+    # FIXME(jathan): As one example, `DeviceFilterSet.q` filter searches in `comments`. The issue
+    # really being that this field renders as a textarea and it's not cute in the UI. Might be able
+    # to dynamically change the widget if we decide we do want to support this field.
+    #
+    # Type: tuple
+    exclude_filter_fields = ("q", "cr", "cf", "comments")  # Must be a tuple
+
+    @classproperty
+    def base_url(cls):
+        if cls.model is None:
+            return
```

Comment:
Yeah I think this is another thing from before the major filterset refactor.

Reply:
Actually, now that I ripped it out, here is why:

```python
In [10]: dg = DynamicGroup()
...
AttributeError: 'NoneType' object has no attribute '_meta'
```

It's because `base_url` is a `@classproperty` so it gets evaluated at access time.